### DISCUSSION TITLE
Best way to remove a number of elements from a JsonArray

### DISCUSSION THREAD

**Discussion by jasonmalinowski** [dotnet/runtime#106726] [Q&A] [answered]
We have a piece of code we're trying to write that is parses JSON into a JsonArray, and needs to trim off a bunch of entries from the JsonArray at the end. For example, the JsonArray might a few hundred entries, but we want to trim it down to just 10.

What's the fastest way to do this? Ideally JsonArray would have a RemoveRange method but it doesn't. JsonArray has a RemoveAt which we can call in a loop but that seems a bit silly. We saw there's a ReplaceWith() method, and the hope was we can do something like:

    array.ReplaceWith(new JsonArray(array.Take(10)))

But that creation of a new JsonArray fails since the nodes we're trying to add already have the old array as a parent. It seems either you have to DeepClone() your entries in the array (ick!) or only replace arrays with empty arrays. :smile:

(I'm also suspicious of ReplaceWith() in general since if you're trying to replace the root node it looks like it'll just [immediately exit](https://github.com/dotnet/runtime/blame/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonNode.cs#L322) which might be a bug?)

Is there some good way to do this? Otherwise is there any reason we can't add a RemoveRange() method, or a method on JsonArray that rather than accepting a JsonArray accepts an array of elements to replace?

Tagging @dotnet/area-system-text-json and @fanyang-mono.

**Comment by steveberdy** [comment]
JsonArray implements `IEnumerable<JsonNode?>`. If I were you, I would try creating a new enumerable that yields the nodes you want.

**Reply by jasonmalinowski** [reply]
But ultimately I still want it as an JsonArray to convert back to text.

**Reply by steveberdy** [reply]
Here's my code example. I hope it helps. You want to clear the original array before constructing a new JsonArray with the new nodes. Otherwise, the code gives the same error you have.
```csharp
public static void Filter()
{
    JsonArray originalArray = (JsonArray)JsonNode.Parse("[1, 2, 3]");
    // Filter out 1's in the JSON array
    static IEnumerable<JsonNode> Generate(JsonArray arr)
    {
        for (var i = 0; i < arr.Count; i++)
        {
            if (arr[i].GetValue<int>() != 1)
            {
                yield return arr[i];
            }
        }
    }
    // Make sure to enumerate through the generator or the array won't be generated
    // before clearing the original array on the next line, resulting in an empty array
    var res = Generate(originalArray).ToArray();
    originalArray.Clear();
    JsonArray s = new(res);
    Console.WriteLine(s.ToString());
}
```

**Reply by steveberdy** [reply]
Tuned to your context
```csharp
JsonArray array = ...;
var nodes = array.Take(10).ToArray();
array.Clear();
var newArray = new JsonArray(nodes);
```

**Reply by CyrusNajmabadi** [reply]
#jasonsagainstjson

**Comment by eiriktsarpalis** [comment]
> Otherwise is there any reason we can't add a RemoveRange() method, or a method on JsonArray that rather than accepting a JsonArray accepts an array of elements to replace?

No reason that I'm aware of. Would you like to file an API proposal?

**Reply by steveberdy** [reply]
This would be cool. Also, it might be better to clarify in documentation that clearing the original array would help ensure the new array doesn't have any issues with the nodes already having parents.

**Reply by jasonmalinowski** [reply, marked as answer]
I filed an API proposal, and then forgot to link it here: https://github.com/dotnet/runtime/issues/107962
